pilot/pkg/networking/plugin/authn/authentication.go

```diff
@@ -41,9 +50,114 @@ func NewPlugin() *Plugin {
 	return &Plugin{}
 }
 
+// RequireTLS returns true and pointer to mTLS params if the policy use mTLS for (peer) authentication.
+// (note that mTLS params can still be nil). Otherwise, return (false, nil).
+func RequireTLS(policy *authn.Policy) (bool, *authn.MutualTls) {
+	if policy == nil {
+		return false, nil
+	}
+	if len(policy.Peers) > 0 {
+		for _, method := range policy.Peers {
+			switch method.GetParams().(type) {
+			case *authn.PeerAuthenticationMethod_Mtls:
+				return true, method.GetMtls()
+			default:
+				continue
+			}
+		}
+	}
+	return false, nil
+}
+
+// JwksURIClusterName returns cluster name for the jwks URI. This should be used
+// to override the name for outbound cluster that are added for Jwks URI so that they
+// can be referred correctly in the JWT filter config.
+func JwksURIClusterName(hostname string, port *model.Port) string {
+	const clusterPrefix = "jwks."
+	const maxClusterNameLength = 189 - len(clusterPrefix)
+	name := hostname + "|" + port.Name
+	if len(name) > maxClusterNameLength {
+		prefix := name[:maxClusterNameLength-sha1.Size*2]
+		sum := sha1.Sum([]byte(name))
+		name = fmt.Sprintf("%s%x", prefix, sum)
+	}
+	return clusterPrefix + name
+}
+
+// CollectJwtSpecs returns a list of all JWT specs (ponters) defined the policy. This
+// provides a convenient way to iterate all Jwt specs.
+func CollectJwtSpecs(policy *authn.Policy) []*authn.Jwt {
+	ret := []*authn.Jwt{}
+	if policy == nil {
+		return ret
+	}
+	for _, method := range policy.Peers {
+		switch method.GetParams().(type) {
+		case *authn.PeerAuthenticationMethod_Jwt:
+			ret = append(ret, method.GetJwt())
+		}
+	}
+	for _, method := range policy.Origins {
+		ret = append(ret, method.Jwt)
+	}
+	return ret
+}
+
+// OutputLocationForJwtIssuer returns the header location that should be used to output payload if
+// authentication succeeds.
+func OutputLocationForJwtIssuer(issuer string) string {
+	const locationPrefix = "istio-sec-"
+	sum := sha1.Sum([]byte(issuer))
```

Comment:
can we just use issuer directly?  such as "sec-istio-" + issuer.   it will be easier to debug

Reply:
true, but we need some sanitization schema. We can chose how to do it later.